### PR TITLE
feat: add ESM support

### LINKED ISSUE #9: Move to ESM
Hello,

a few days ago Node.js 10 reached it's EOL so it might be worth converting this dependency to ES Modules ([node.js docs](https://nodejs.org/api/esm.html#esm_introduction)). LTS 12.x and 14.x will still work.

> ECMAScript modules are the official standard format to package JavaScript code for reuse. Modules are defined using a variety of import and export statements.

Step-by-step guide of migrating to ESM is here: https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

### PR DESCRIPTION
**Pull request by talentlessguy**
closes #9

## What is the motivation for this pull request?

Support ESM (and browser support as well)

## What is the current behavior?

Because of it's being CJS it doesn't work out-of-the-box in browsers. With ESM it will not only support Node >=12 but also unpkg.com

## What is the new behavior?

it supports esm :)

### CODE REVIEW

File: index.mjs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import mod from './index.js';
```
Comment (remarkablemark) [inline index.mjs:1]: Can we rename `mod` to `StyleToObject` for explicitness?

File: index.mjs [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+
+export const StyleToObject = mod;
```
Comment (remarkablemark) [inline index.mjs:3]: Should we be consistent with how the package does CommonJS export here?

```js
export default mod;
```

File: package.json [hunk 3]
```diff
@@ -5,6 +5,15 @@
   "author": "Mark <mark@remarkablemark.org>",
   "main": "index.js",
   "types": "index.d.ts",
+  "module": "index.mjs",
+  "exports": {
```
Comment (remarkablemark) [inline package.json:9]: Is `exports` necessary? See [html-react-parser](https://github.com/remarkablemark/html-react-parser/commit/ca0340a25e9c9c56758656b3352267ade7e37a21#diff-d6299b5bb2afc0956126a0adb12fd89829e01a4cc025e7277a092228b96f32e4)

**Review by remarkablemark — CHANGES_REQUESTED** [review]
Thanks for opening this PR @talentlessguy.

I left some comments and can we add [integration tests](https://github.com/remarkablemark/html-react-parser/commit/2659967923dad16c6445e9a64b7ed184051a9b5e) for this?

Also, do you mind updating your commit messages with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)? It's used to help with release and changelog generation:

```sh
git commit --amend -m 'feat: add esm support'
git push -f
```

File: package.json (hunk 3 above)
Comment (talentlessguy) [reply package.json:9]: Node.js ignores "module" field and only reads from "main" / "exports". that's why we need it there

doc: https://nodejs.org/api/packages.html#packages_exports

File: index.mjs (hunk 2 above)
Comment (talentlessguy) [reply index.mjs:3]: ok no prob, will replace with default export

File: package.json (hunk 3 above)
Comment (remarkablemark) [reply package.json:9]: Gotcha, thanks for sending the doc. Can we add an example or a test to verify that the field is necessary here?

**Comment by talentlessguy** [comment]
@remarkablemark hi, looks I've completed all of your tasks

- tests are present
- "main" and "module" field are present as well

**Comment by remarkablemark** [comment]
> @remarkablemark hi, looks I've completed all of your tasks
> 
> * tests are present
> * "main" and "module" field are present as well

Thanks I will take a look

File: test/esm.mjs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import assert from 'assert'
+import StyleToObject from '../index.mjs'
+
+describe('it works with ESM', () => {
+	it('imports the functiion without errors', () => {
```
Comment (remarkablemark) [inline test/esm.mjs:5]: Spelling error `s/functiion/function/`

Comment (remarkablemark) [inline test/esm.mjs:5]: Nit: we prefer 2 spaces over 1 tab

**Review by remarkablemark — CHANGES_REQUESTED** [review]
Thanks @talentlessguy for making the changes, I left a few comments.

**Comment by remarkablemark** [comment]
Thanks for following [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) in your recent commit messages. However, [commitlint](https://github.com/conventional-changelog/commitlint/#what-is-commitlint) is complaining about the commit messages that weren't updated:

```
Run npx commitlint --from=HEAD~1
⧗   input: add "module" entrypoint (for browsers)
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint

⧗   input: add esm support
✖   subject may not be empty [subject-empty]
✖   type may not be empty [type-empty]

✖   found 2 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
```

Feel free to amend those commit messages and force push.

File: test/esm.mjs (hunk 4 above)
Comment (talentlessguy) [reply test/esm.mjs:5]: @remarkablemark oops sorry will fix

btw i recommend setting up git hooks to run formatter and linter before committing

File: test/esm.mjs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import assert from 'assert'
```
[the remark this answers is not in the document]
Comment (talentlessguy) [reply test/esm.mjs]: @remarkablemark looks like already runs on `npm test`?

```
➜ pnpm test

> style-to-object@0.3.0 test /home/v1rtl/Coding/forks/style-to-object
> mocha



  it works with ESM
    ✓ imports the function without errors
# blah blah
```

**Review by remarkablemark — APPROVED** [review]
LGTM, thanks for working on this @talentlessguy!

**Comment by talentlessguy** [comment]
@remarkablemark squashing the commits will join them into one

**Comment by talentlessguy** [comment]
@remarkablemark hello, I've rebased some of the commits in this PR, now they all follow the Commitlint spec

**Review by remarkablemark — CHANGES_REQUESTED** [review]
@talentlessguy thanks for rebasing the commits to follow Conventional Commits. However, it seems that some extraneous commits got into this PR (see [Files changed](https://github.com/remarkablemark/style-to-object/pull/10/files)).

Do you mind rebasing again and dropping the extraneous commits?

**Comment by remarkablemark** [comment]
@talentlessguy could you help resolve merge conflicts?

**Comment by remarkablemark** [comment]
@talentlessguy I appreciate your original work on adding ESM support. I ended up opening #21 to close the loop. Thank you once again for your effort and my apologies for forgetting this PR and not following up sooner.

**Pull request closed without merging**